File: rust/prob2.rs
fn fib(n: int) -> int {
    fn _fib(n: int, a: int, b: int) -> int {
        match (n, a, b) {
            (0, _,  _) => a,
            _          => _fib(n - 1, a + b, a)
        }
    }

    _fib(n, 0, 1)
}

fn main() {
    let mut fibat = 1;
    let mut fibn = 0;
    let mut sum = 0;
    while fibn <= 4000000 {
        fibn = fib(fibat);
        fibat += 1;
        if fibn % 2 == 0 {
            sum += fibn
        };
    } 
    println!("{:i}", sum);
}





fn euler2(n: int, m: int, limit: int, total: int) -> int {
    match (n, m, limit, total) {
        (_, m, limit, total) if m > limit => total,
        (n, m, limit, total) => {
            match (n+m) % 2 {
                0 => euler2(m, n+m, limit, total+n+m),
                1 => euler2(m, n+m, limit, total),
                _ => { println!("Whoops"); 0 }
            }
        }
    }

}


fn main() {
    let sum = euler2(1, 2, 4000000, 2);
    println!("{:i}", sum);
}

File: clojure/prob2.clj
(defn fib [n]
  "Caclculates the n-th fibbonacci number"
  (defn _fib [n, a, b]
    (if (== n 0) 
      a
      (_fib (- n 1) (+ a b) a)))
  (_fib n 0 1))

(let
    [fibat (ref 1)
     fibn  (ref 0)
     sum   (ref 0)]
  (while (<= @fibn 4000000)
    (dosync
     (ref-set fibn (fib @fibat))
     (alter fibat + 1)
     (if (== 0 (mod @fibn 2))
       (alter sum + @fibn))))
  (println @sum))
